src/cdomains/apron/linearTwoVarEqualityDomain.apron.ml

```diff
@@ -0,0 +1,927 @@
+         | (None, b) ->
+           (match ts'.(i) with
+            | (None, b') -> if not @@ Z.equal b b' then ts := None;
+            | (Some j, b') -> subst_var ts j (None, Z.(b - b')))
+         | (Some j, b) ->
+           (match ts'.(i) with
+            | (None, b1) -> subst_var ts j (None, Z.(b1 - b))
+            | (Some h1, b1) ->
+              (match ts'.(j) with
+               | (None, b2) -> subst_var ts i (None, Z.(b2 + b))
+               | (Some h2, b2) ->
+                 if h1 = h2 then
+                   (if Z.(b1 <> (b2 + b)) then ts := None)
+                 else if h1 < h2 then subst_var ts h2 (Some h1, Z.(b1 - (b + b2)))
+                 else subst_var ts h1 (Some h2, Z.(b + (b2 - b1))))))
+      in
+      Stdlib.Option.iter adjust !ts
+    in
+    match t1.d, t2.d with
+    | Some d1', Some d2' -> (
+        let ds = ref (Some (Array.copy d1')) in
+        Array.iteri (fun j e -> add_conj ds e j) d2';
+        {d = !ds; env = sup_env} )
+    | _ -> { d = None; env = sup_env}
+
+  let meet t1 t2 =
+    let res = meet t1 t2 in
+    if M.tracing then M.tracel "meet" "meet a: %s b: %s -> %s \n" (show t1) (show t2) (show res) ;
+    res
+
+  let meet t1 t2 = timing_wrap "meet" (meet t1) t2
+
+  let leq t1 t2 =
+    let env_comp = Environment.compare t1.env t2.env in (* Apron's Environment.compare has defined return values. *)
+    let implies ts t i : bool =
+      match t with
+      | (None, b) ->
+        (match ts.(i) with
+         | (None, b') -> Z.equal b b'
+         | _ -> false)
+      | (Some j, b) ->
+        (match ts.(i), ts.(j) with
+         | (None, b1), (None, b2) -> Z.equal b1 (Z.add b2 b)
+         | (Some h1, b1), (Some h2, b2) ->
+           h1 = h2 && Z.equal b1 (Z.add b2 b)
+         | _ -> false
+        )
+    in
+    if env_comp = -2 || env_comp > 0 then false else
+    if is_bot_env t1 || is_top t2 then true else
+    if is_bot_env t2 || is_top t1 then false else
+      let m1, m2 = Option.get t1.d, Option.get t2.d in
+      let m1' = if env_comp = 0 then m1 else dim_add (Environment.dimchange t1.env t2.env) m1 in
+      GobArray.for_alli (fun i t -> implies m1' t i) m2
+
+  let leq a b = timing_wrap "leq" (leq a) b
+
+  let leq t1 t2 =
+    let res = leq t1 t2 in
+    if M.tracing then M.tracel "leq" "leq a: %s b: %s -> %b \n" (show t1) (show t2) res ;
+    res
+
+  let join a b =
+    let join_d ad bd =
+      (*use copy of ad because result is later saved in there*)
+      let ad = Array.copy ad in
+      (*This is the table which is later grouped*)
+      let table = BatList.map2i (fun i a b -> (i, a, b)) (Array.to_list ad) (Array.to_list bd) in
+      let const_offset t = Tuple2.second t in
+      let diff t1 t2 = Z.(const_offset t1 - const_offset t2) in
+      (*compare two variables for grouping depending on delta function and reference index*)
+      let cmp_z (_, t1i, t2i) (_, t1j, t2j) =
+        Tuple3.compare (fst t1i, fst t2i, diff t1i t2i) (fst t1j, fst t2j, diff t1j t2j)
```

Comment:
Nice find with the Tuple3.compare! It is indeed a compare that goes element-by-element from left to right.

It internally uses `Stdlib.compare` (https://v2.ocaml.org/releases/5.1/api/Stdlib.html), i.e., polymorphic compare, for the elements of the tuples which gives an order compatible with `(>=)`

> Structural ordering functions. These functions coincide with the usual orderings over integers, characters, strings, byte sequences and floating-point numbers, and extend them to a total ordering over all types. 

Here's a bit of annoying detail: Here we rely on a special order for `int option` that ensures that `None < Some _` which probably is true in the implementation, but not guaranteed (the standard just says any total order). Same goes for the last component which is of type `Z.t` where polymorphic comparison is not guaranteed to give the correct result at all (equivalent values may in theory be represented differently depending on whether they are represented directly or through the C binding).


So using the default polymorphic compare here wouldn't cut it. Luckily Tuple3 lets us specify the comparison function.

```suggestion
        let cmp_ref = Option.compare Int.compare in
        Tuple3.compare ~cmp1:cmp_ref ~cmp2:cmp_ref ~cmp3:Z.compare (fst t1i, fst t2i, diff t1i t2i) (fst t1j, fst t2j, diff t1j t2j)
```

Reply:
thanks for the suggestion 👍

Reply:
Ah, I guess `Batteries` and `Stdlib` are incompatible here, which is why you needed the `~cmp:` thing you added.
This is really an annoying thing that batteries and stdlib now contradict each other :/

Reply:
@sim642 Do you know what the batteries folks are aiming for? Compatibility with Stdlib or backwards-compatibility? Might be worth making an issue if there's chance to get it fixed.

Reply:
At this point backwards-compatibility more likely. There's no real fix to this because I guess batteries added `Option.compare` before and stdlib got it later (but in a different signature), so it's impossible to remain compatible with both. This is also apparent in gems like `BatString.starts_with_stdlib`.

In this case, I'd just use deriving for the whole thing:
```ocaml
[%ord: int option * int option * Z.t]
```
instead of the `Tuple3.compare` etc.

Reply:
The algorithm depends on a specific order for correctness though, it's not like any total order will do. Does the deriver give any guarantees about the order or derives other than it is total?